Add device logic handling for functions which allow scalar inputs as tensors

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86149
* #86144

Some functions allow scalars as tensor inputs. Add handling for them in device logic.

Fix for https://github.com/pytorch/torchdynamo/issues/1445

eellison committed:
Add device logic handling for functions which allow scalar inputs as tensors

[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -567,6 +567,14 @@ def merge_devices(t):
+        # if we have failed to find a device, and we're running one of these operators,
+        # we must have scalar only inputs
+        if torch._C._should_allow_numbers_as_tensors(
+            func.name().split("::")[-1].split(".")[0]
```
I feel that func already has an API for getting this directly, and if it does not, you should add one.

eellison replied:
`unqualified_name` I guess ?

eellison committed:
Update on "Add device logic handling for functions which allow scalar inputs as tensors"


Some functions allow scalars as tensor inputs. Add handling for them in device logic.

Fix for https://github.com/pytorch/torchdynamo/issues/1445

[ghstack-poisoned]

eellison committed:
Update on "Add device logic handling for functions which allow scalar inputs as tensors"


Some functions allow scalars as tensor inputs. Add handling for them in device logic.

Fix for https://github.com/pytorch/torchdynamo/issues/1445

[ghstack-poisoned]

eellison committed:
Update on "Add device logic handling for functions which allow scalar inputs as tensors"


Some functions allow scalars as tensor inputs. Add handling for them in device logic.

Fix for https://github.com/pytorch/torchdynamo/issues/1445

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge -f "dont need slow tests"

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot rebase

eellison committed:
Update on "Add device logic handling for functions which allow scalar inputs as tensors"


Some functions allow scalars as tensor inputs. Add handling for them in device logic.

Fix for https://github.com/pytorch/torchdynamo/issues/1445

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
